Notebooks/SentinelUtilities/SentinelPortal/bookmark_helper.py

```diff
@@ -0,0 +1,318 @@
+from SentinelUtils import InputValidation
+
+
+class Constants():
+    """ This class holds constants """
+    TYPE = 'Microsoft.SecurityInsights/Bookmarks'
+    ETAG = '*'
+    BOOKMARK_RESOURCE_BASE = '/subscriptions/{0}/resourceGroups/{1}/providers/Microsoft.OperationalInsights/workspaces/{2}/providers/Microsoft.SecurityInsights/bookmarks'
+    BOOKMARK_ID = '/{0}'
+    BOOKMARK_RESOURCE_ID = BOOKMARK_RESOURCE_BASE + BOOKMARK_ID
+
+class EntityType():
+    """ This class holds entity type enums """
+    BOOKMARK = 'Bookmark'
+    SECURITYALERT = 'SecurityAlert'
+    ACCOUNT = 'Account'
+    HOST = 'Host'
+    MALWARE = 'Malware'
+    FILE = 'File'
+    PROCESS = 'Process'
+    CLOUDAPPLICATION = 'CloudApplication'
+    DNSRESOLUTION = 'DnsResolution'
+    AZURERESOURCE = 'AzureResource'
+    FILEHASH = 'FileHash'
+    REGISTRYKEY = 'RegistryKey'
+    REGISTRYVALUE = 'RegistryValue'
+    SECURITYGROUP = 'SecurityGroup'
+    URL = 'Url'
+
+class BookmarkProperties():
+    """ This class holds properties for Hunting bookmark """
+
+    def __init__(self,
+                 display_name,
+                 query,
+                 query_result=None,
+                 tag_list=None,
+                 notes=None,
+                 event_time=None,
+                 query_start_time=None,
+                 query_end_time=None):
+        self.displayName = display_name
+        self.labels = tag_list
+        self.query = query
+        self.queryResult = query_result
+        self.notes = notes
+        self.eventTime = event_time
+        self.queryStartTime = query_start_time
+        self.queryEndTime = query_end_time
+        self.bookmarkId = str(uuid.uuid4())
+
+    @property
+    def bookmarkId(self):
+        """ bookmarkId getter """
+        return self._bookmarkId
+
+    @bookmarkId.setter
+    def bookmarkId(self, bookmark_id):
+        """ bookmarkId setter """
+        self._bookmarkId = bookmark_id
+
+    @property
+    def displayName(self):
+        """ displayName getter """
+        return self._displayName
+
+    @displayName.setter
+    def displayName(self, display_name):
+        """ displayName setter """
+        InputValidation.validate_input('display_name', display_name)
+        self._displayName = display_name
+
+    @property
+    def labels(self):
+        """ labels getter """
+        return self._labels
+
+    @labels.setter
+    def labels(self, tag_list):
+        """ labels setter """
+        if tag_list:
+            self._labels = tag_list
+        else:
+            self._labels = []
+
+    @property
+    def query(self):
+        """ query """
+        return self._query
+
+    @query.setter
+    def query(self, query):
+        """ query setter """
+        InputValidation.validate_input('query', query)
+        self._query = query
+
+    @property
+    def queryResult(self):
+        """ queryResult """
+        return self._queryResult
+
+    @queryResult.setter
+    def queryResult(self, query_result):
```

Comment:
query_result is a complex object itself - a dict with an _entityMapping property that is another dict. Do you think it would be worth creating a special query result object that breaks these to make it easier for people to create. E.g. qr = QueryResult(data=my_data, entity_map=my_mapping)

Reply:
Agreed, good point.  Here I will let users pass in dictionary, then I will process the passed in value in setter.